### ISSUE TITLE
Installation fails (Python 3.10.2)

### ISSUE THREAD

**Issue by davnovak** [saeyslab/FlowSOM_Python#10]
### Report

Installation fails on an arm64 Mac conda environment with Python 3.10.2:

<details>
<summary>Session info</summary>

```
-----
session_info        1.0.0
-----
Python 3.10.2 | packaged by conda-forge | (main, Mar  8 2022, 15:52:47) [Clang 11.1.0 ]
macOS-14.1.1-arm64-arm-64bit
-----
Session information updated at 2024-09-06 14:38
```

</details>

I think this is new; around two months ago there was no problem.
Environment creation and activation runs fine: `conda create -n flowsom -y python==3.10.2; conda activate flowsom`
But then, installing FlowSOM fails at `statsmodels-0.13.0`, required by `scanpy` (see _Logs_ for output).

<details>
  <summary>
  Logs
  </summary>

  ```zsh
(flowsom) user@machine ~ % pip install git+https://github.com/saeyslab/FlowSOM_Python.git
Collecting git+https://github.com/saeyslab/FlowSOM_Python.git
  Cloning https://github.com/saeyslab/FlowSOM_Python.git to /private/var/folders/dq/kfj02tc94z1f_z3jfq1y_8p80000gn/T/pip-req-build-r78ggdjl
  Running command git clone --filter=blob:none --quiet https://github.com/saeyslab/FlowSOM_Python.git /private/var/folders/dq/kfj02tc94z1f_z3jfq1y_8p80000gn/T/pip-req-build-r78ggdjl
  Resolved https://github.com/saeyslab/FlowSOM_Python.git to commit 80529c6b7a1747e8e71042102ac8762c3bfbaa1b
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting anndata (from flowsom==0.0.1)
  Using cached anndata-0.10.9-py3-none-any.whl.metadata (6.9 kB)
Collecting igraph (from flowsom==0.0.1)
  Using cached igraph-0.11.6-cp39-abi3-macosx_11_0_arm64.whl.metadata (3.9 kB)
Collecting loguru (from flowsom==0.0.1)
  Using cached loguru-0.7.2-py3-none-any.whl.metadata (23 kB)
Requirement already satisfied: matplotlib in ./.local/lib/python3.10/site-packages/matplotlib-3.6.0rc1-py3.10-macosx-11.0-arm64.egg (from flowsom==0.0.1) (3.6.0rc1)
Collecting mudata (from flowsom==0.0.1)
  Using cached mudata-0.3.1-py3-none-any.whl.metadata (8.3 kB)
Collecting numba (from flowsom==0.0.1)
  Downloading numba-0.60.0-cp310-cp310-macosx_11_0_arm64.whl.metadata (2.7 kB)
Collecting numpy (from flowsom==0.0.1)
  Downloading numpy-2.1.1-cp310-cp310-macosx_14_0_arm64.whl.metadata (60 kB)
Collecting pandas (from flowsom==0.0.1)
  Downloading pandas-2.2.2-cp310-cp310-macosx_11_0_arm64.whl.metadata (19 kB)
Collecting pytometry (from flowsom==0.0.1)
  Using cached pytometry-0.1.5-py3-none-any.whl.metadata (3.0 kB)
Collecting readfcs (from flowsom==0.0.1)
  Using cached readfcs-1.1.8-py2.py3-none-any.whl.metadata (1.9 kB)
Collecting scanpy (from flowsom==0.0.1)
  Using cached scanpy-1.10.2-py3-none-any.whl.metadata (9.3 kB)
Collecting scikit-learn (from flowsom==0.0.1)
  Downloading scikit_learn-1.5.1-cp310-cp310-macosx_12_0_arm64.whl.metadata (12 kB)
Collecting scipy (from flowsom==0.0.1)
  Downloading scipy-1.14.1-cp310-cp310-macosx_14_0_arm64.whl.metadata (60 kB)
Collecting seaborn (from flowsom==0.0.1)
  Using cached seaborn-0.13.2-py3-none-any.whl.metadata (5.4 kB)
Collecting session-info (from flowsom==0.0.1)
  Using cached session_info-1.0.0.tar.gz (24 kB)
  Preparing metadata (setup.py) ... done
Collecting array-api-compat!=1.5,>1.4 (from anndata->flowsom==0.0.1)
  Using cached array_api_compat-1.8-py3-none-any.whl.metadata (1.5 kB)
Collecting exceptiongroup (from anndata->flowsom==0.0.1)
  Downloading exceptiongroup-1.2.2-py3-none-any.whl.metadata (6.6 kB)
Collecting h5py>=3.1 (from anndata->flowsom==0.0.1)
  Downloading h5py-3.11.0-cp310-cp310-macosx_11_0_arm64.whl.metadata (2.5 kB)
Collecting natsort (from anndata->flowsom==0.0.1)
  Using cached natsort-8.4.0-py3-none-any.whl.metadata (21 kB)
Collecting packaging>=20.0 (from anndata->flowsom==0.0.1)
  Using cached packaging-24.1-py3-none-any.whl.metadata (3.2 kB)
Collecting python-dateutil>=2.8.2 (from pandas->flowsom==0.0.1)
  Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)
Collecting pytz>=2020.1 (from pandas->flowsom==0.0.1)
  Using cached pytz-2024.1-py2.py3-none-any.whl.metadata (22 kB)
Collecting tzdata>=2022.7 (from pandas->flowsom==0.0.1)
  Using cached tzdata-2024.1-py2.py3-none-any.whl.metadata (1.4 kB)
Collecting texttable>=1.6.2 (from igraph->flowsom==0.0.1)
  Using cached texttable-1.7.0-py2.py3-none-any.whl.metadata (9.8 kB)
Collecting contourpy>=1.0.1 (from matplotlib->flowsom==0.0.1)
  Downloading contourpy-1.3.0-cp310-cp310-macosx_11_0_arm64.whl.metadata (5.4 kB)
Collecting cycler>=0.10 (from matplotlib->flowsom==0.0.1)
  Using cached cycler-0.12.1-py3-none-any.whl.metadata (3.8 kB)
Collecting fonttools>=4.22.0 (from matplotlib->flowsom==0.0.1)
  Downloading fonttools-4.53.1-cp310-cp310-macosx_11_0_arm64.whl.metadata (162 kB)
Collecting kiwisolver>=1.0.1 (from matplotlib->flowsom==0.0.1)
  Downloading kiwisolver-1.4.7-cp310-cp310-macosx_11_0_arm64.whl.metadata (6.3 kB)
Collecting pillow>=6.2.0 (from matplotlib->flowsom==0.0.1)
  Downloading pillow-10.4.0-cp310-cp310-macosx_11_0_arm64.whl.metadata (9.2 kB)
Collecting pyparsing>=2.2.1 (from matplotlib->flowsom==0.0.1)
  Using cached pyparsing-3.1.4-py3-none-any.whl.metadata (5.1 kB)
Collecting llvmlite<0.44,>=0.43.0dev0 (from numba->flowsom==0.0.1)
  Downloading llvmlite-0.43.0-cp310-cp310-macosx_11_0_arm64.whl.metadata (4.8 kB)
Collecting numpy (from flowsom==0.0.1)
  Downloading numpy-2.0.2-cp310-cp310-macosx_14_0_arm64.whl.metadata (60 kB)
Collecting nbproject (from pytometry->flowsom==0.0.1)
  Using cached nbproject-0.10.4-py2.py3-none-any.whl.metadata (2.7 kB)
Collecting pandas (from flowsom==0.0.1)
  Downloading pandas-1.5.3-cp310-cp310-macosx_11_0_arm64.whl.metadata (11 kB)
Collecting flowutils (from pytometry->flowsom==0.0.1)
  Downloading FlowUtils-1.1.0-cp310-cp310-macosx_11_0_arm64.whl.metadata (3.0 kB)
Collecting datashader (from pytometry->flowsom==0.0.1)
  Using cached datashader-0.16.3-py2.py3-none-any.whl.metadata (12 kB)
Collecting consensusclustering (from pytometry->flowsom==0.0.1)
  Using cached consensusclustering-0.2.3-py3-none-any.whl.metadata (2.0 kB)
Collecting minisom (from pytometry->flowsom==0.0.1)
  Using cached MiniSom-2.3.3.tar.gz (11 kB)
  Preparing metadata (setup.py) ... done
Collecting lamin_utils (from readfcs->flowsom==0.0.1)
  Using cached lamin_utils-0.13.4-py2.py3-none-any.whl.metadata (980 bytes)
Collecting fcsparser (from readfcs->flowsom==0.0.1)
  Using cached fcsparser-0.2.8-py3-none-any.whl.metadata (3.6 kB)
Collecting joblib (from scanpy->flowsom==0.0.1)
  Using cached joblib-1.4.2-py3-none-any.whl.metadata (5.4 kB)
Collecting legacy-api-wrap>=1.4 (from scanpy->flowsom==0.0.1)
  Using cached legacy_api_wrap-1.4-py3-none-any.whl.metadata (1.8 kB)
Collecting matplotlib (from flowsom==0.0.1)
  Downloading matplotlib-3.9.2-cp310-cp310-macosx_11_0_arm64.whl.metadata (11 kB)
Collecting networkx>=2.7 (from scanpy->flowsom==0.0.1)
  Using cached networkx-3.3-py3-none-any.whl.metadata (5.1 kB)
Collecting numpy (from flowsom==0.0.1)
  Downloading numpy-1.26.4-cp310-cp310-macosx_11_0_arm64.whl.metadata (61 kB)
Collecting patsy (from scanpy->flowsom==0.0.1)
  Using cached patsy-0.5.6-py2.py3-none-any.whl.metadata (3.5 kB)
Collecting pynndescent>=0.5 (from scanpy->flowsom==0.0.1)
  Using cached pynndescent-0.5.13-py3-none-any.whl.metadata (6.8 kB)
Collecting statsmodels>=0.13 (from scanpy->flowsom==0.0.1)
  Downloading statsmodels-0.14.2-cp310-cp310-macosx_11_0_arm64.whl.metadata (9.2 kB)
Collecting tqdm (from scanpy->flowsom==0.0.1)
  Using cached tqdm-4.66.5-py3-none-any.whl.metadata (57 kB)
Collecting umap-learn!=0.5.0,>=0.5 (from scanpy->flowsom==0.0.1)
  Using cached umap_learn-0.5.6-py3-none-any.whl.metadata (21 kB)
Collecting threadpoolctl>=3.1.0 (from scikit-learn->flowsom==0.0.1)
  Using cached threadpoolctl-3.5.0-py3-none-any.whl.metadata (13 kB)
Collecting stdlib_list (from session-info->flowsom==0.0.1)
  Using cached stdlib_list-0.10.0-py3-none-any.whl.metadata (3.3 kB)
Collecting six>=1.5 (from python-dateutil>=2.8.2->pandas->flowsom==0.0.1)
  Using cached six-1.16.0-py2.py3-none-any.whl.metadata (1.8 kB)
Collecting kneed<0.9.0,>=0.8.5 (from consensusclustering->pytometry->flowsom==0.0.1)
  Using cached kneed-0.8.5-py3-none-any.whl.metadata (5.5 kB)
INFO: pip is looking at multiple versions of consensusclustering to determine which version is compatible with other requirements. This could take a while.
Collecting consensusclustering (from pytometry->flowsom==0.0.1)
  Using cached consensusclustering-0.2.1-py3-none-any.whl.metadata (2.1 kB)
Collecting statsmodels>=0.13 (from scanpy->flowsom==0.0.1)
  Downloading statsmodels-0.14.1-cp310-cp310-macosx_11_0_arm64.whl.metadata (9.5 kB)
  Downloading statsmodels-0.14.0-cp310-cp310-macosx_11_0_arm64.whl.metadata (9.0 kB)
  Downloading statsmodels-0.13.5-cp310-cp310-macosx_11_0_arm64.whl.metadata (9.5 kB)
INFO: pip is still looking at multiple versions of consensusclustering to determine which version is compatible with other requirements. This could take a while.
  Downloading statsmodels-0.13.4-cp310-cp310-macosx_11_0_arm64.whl.metadata (9.5 kB)
  Downloading statsmodels-0.13.3-cp310-cp310-macosx_11_0_arm64.whl.metadata (9.5 kB)
  Downloading statsmodels-0.13.2-cp310-cp310-macosx_11_0_arm64.whl.metadata (8.1 kB)
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. See https://pip.pypa.io/warnings/backtracking for guidance. If you want to abort this run, press Ctrl + C.
  Downloading statsmodels-0.13.1-cp310-cp310-macosx_11_0_arm64.whl.metadata (8.0 kB)
  Downloading statsmodels-0.13.0.tar.gz (17.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.8/17.8 MB 15.5 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [67 lines of output]
      <string>:19: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
      <string>:53: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
      <string>:56: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
      <string>:53: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.

      Error compiling Cython file:
      ------------------------------------------------------------
      ...
      cimport cython
      cimport numpy as np
      import numpy as np

      ctypedef np.float64_t DOUBLE
      ctypedef np.int_t INT
               ^
      ------------------------------------------------------------

      statsmodels/nonparametric/linbin.pyx:13:9: 'int_t' is not a type identifier
      Compiling statsmodels/tsa/_stl.pyx because it changed.
      Compiling statsmodels/tsa/holtwinters/_exponential_smoothers.pyx because it changed.
      Compiling statsmodels/tsa/exponential_smoothing/_ets_smooth.pyx because it changed.
      Compiling statsmodels/tsa/_innovations.pyx because it changed.
      Compiling statsmodels/tsa/regime_switching/_hamilton_filter.pyx because it changed.
      Compiling statsmodels/tsa/regime_switching/_kim_smoother.pyx because it changed.
      Compiling statsmodels/tsa/innovations/_arma_innovations.pyx because it changed.
      Compiling statsmodels/nonparametric/linbin.pyx because it changed.
      Compiling statsmodels/robust/_qn.pyx because it changed.
      Compiling statsmodels/nonparametric/_smoothers_lowess.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_initialization.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_representation.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_kalman_filter.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_filters/_conventional.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_filters/_inversions.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_filters/_univariate.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_filters/_univariate_diffuse.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_kalman_smoother.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_smoothers/_alternative.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_smoothers/_classical.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_smoothers/_conventional.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_smoothers/_univariate.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_smoothers/_univariate_diffuse.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_simulation_smoother.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_cfa_simulation_smoother.pyx because it changed.
      Compiling statsmodels/tsa/statespace/_tools.pyx because it changed.
      [ 1/26] Cythonizing statsmodels/nonparametric/_smoothers_lowess.pyx
      [ 2/26] Cythonizing statsmodels/nonparametric/linbin.pyx
      Traceback (most recent call last):
        File "/Users/davidnovak/miniforge3/envs/flowsom/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/Users/davidnovak/miniforge3/envs/flowsom/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/Users/davidnovak/miniforge3/envs/flowsom/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/private/var/folders/dq/kfj02tc94z1f_z3jfq1y_8p80000gn/T/pip-build-env-3mh6nw2i/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 332, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
        File "/private/var/folders/dq/kfj02tc94z1f_z3jfq1y_8p80000gn/T/pip-build-env-3mh6nw2i/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 302, in _get_build_requires
          self.run_setup()
        File "/private/var/folders/dq/kfj02tc94z1f_z3jfq1y_8p80000gn/T/pip-build-env-3mh6nw2i/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 503, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/private/var/folders/dq/kfj02tc94z1f_z3jfq1y_8p80000gn/T/pip-build-env-3mh6nw2i/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 318, in run_setup
          exec(code, locals())
        File "<string>", line 304, in <module>
        File "/private/var/folders/dq/kfj02tc94z1f_z3jfq1y_8p80000gn/T/pip-build-env-3mh6nw2i/overlay/lib/python3.10/site-packages/Cython/Build/Dependencies.py", line 1154, in cythonize
          cythonize_one(*args)
        File "/private/var/folders/dq/kfj02tc94z1f_z3jfq1y_8p80000gn/T/pip-build-env-3mh6nw2i/overlay/lib/python3.10/site-packages/Cython/Build/Dependencies.py", line 1321, in cythonize_one
          raise CompileError(None, pyx_file)
      Cython.Compiler.Errors.CompileError: statsmodels/nonparametric/linbin.pyx
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

(Some things were cached here, but the problem arises at `statsmodels` regardless.)
</details>

I came up with a work-around for now, but it's one highly redundant Frankenstein's monster of a script:

```zsh
pip install --no-input numpy==1.26.3 numba==0.59.0 matplotlib==3.8.2 scipy==1.12.0 scikit-learn==1.4.0 scanpy==1.9.8 statsmodels pytometry git+https://github.com/saeyslab/FlowSOM_Python.git@80529c6b7a1747e8e71042102ac8762c3bfbaa1b
pip install --no-input "dask[dataframe]" --upgrade
pip install --no-input pandas==2.0.0
```

Cheers

**Comment by davnovak** [comment]
Reproducible in Google Colab as well, so it turns out it is not an Apple Silicon problem

**Comment by LuLeom** [comment]
Thanks David, I have the same problem (python 3.12.5), and managed to fix with you workaround

**Comment by berombau** [comment]
This should be fixed now with pytometry 0.1.6. Can you check again if git install with main works @davnovak or @LuLeom to be sure?

**Comment by davnovak** [comment]
@berombau Thanks, I can confirm it works with my set-up now.
